whetstone.py

```diff
@@ -32,22 +59,482 @@ def _authorize(self):
+            "created",
+            "email",
+            "inactive",
+            "lastActivity",
+            "lastModified",
+            "locked",
+            "name",
+            "first",
+            "last",
+            "school",
+            "course",
+            "coach",
+        ]
+
+    def _preprocess_records(self, records):
+        models = {"Users": []}
+        for record in records:
+            record["school"] = record.get("defaultInformation").get("school")
+            record["course"] = record.get("defaultInformation").get("course")
+            user = {k: v for (k, v) in record.items() if k in self.columns}
+            models["Users"].append(user)
+        return models
+
+
+class Schools(Whetstone):
+    def __init__(self, sql, qa=False):
+        super().__init__(sql)
+        self.columns = [
+            "_id",
+            "internalId",
+            "name",
+            "abbreviation",
+            "archivedAt",
+            "principal",
+            "gradeSpan",
+            "lowGrade",
+            "highGrade",
+            "district",
+            "phone",
+            "address",
+            "city",
+            "cluster",
+            "region",
+            "state",
+            "zip",
+            "lastModified",
+        ]
+
+    def _preprocess_records(self, records):
+        models = {"Schools": [], "ObservationGroups": [], "ObservationGroupMembers": []}
+        for record in records:
+            record_id = record.get("_id")
+            school = {k: v for (k, v) in record.items() if k in self.columns}
+            models["Schools"].append(school)
+            groups = record.get("observationGroups")
+            if groups:
+                for group in groups:
+                    observers = self._extract_group_members(
+                        group, "observers", record_id
+                    )
+                    models["ObservationGroupMembers"].extend(observers)
+                    observees = self._extract_group_members(
+                        group, "observees", record_id
+                    )
+                    models["ObservationGroupMembers"].extend(observees)
+
+                groups = [
+                    {
+                        "id": item.get("_id"),
+                        "name": item.get("name"),
+                        "school": record_id,
+                        "lastModified": item.get("lastModified"),
+                    }
+                    for item in groups
+                ]
+                models["ObservationGroups"].extend(groups)
+        return models
+
+    def _extract_group_members(self, group, role, record_id):
+        """Helper function for formatting ObservationGroupMember records."""
+        members = [
+            dict(
+                item,
+                school=record_id,
+                role=role[:-1],
+                observationGroup=group.get("_id"),
+            )
+            for item in group.get(role)
+        ]
```

Comment:
Similarly here, the multiline line list comprehension is pretty uncommon. Also, why use the `dict` function instead of `{}` syntax?

(This comment / question applies to all the ones below as well)

Reply:
The `dict` function is being used as a simplified constructor. Rather than explicitly building out the dict key by key, I can pass in the existing dict from the endpoint and then append the additional columns needed. I suppose this could also be done with dictionary unpacking but that seemed more complicated and need to be done in a loop. My hope is to abstract this pattern out later as a reusable function do to it's extensibility.